api/src/routes/donate.ts

```diff
@@ -10,71 +15,209 @@ import {
+            return {
+              type: 'info',
+              message: 'User is already donating.'
+            } as const;
+          }
+
+          await fastify.prisma.user.update({
+            where: { id: req.session.user.id },
+            data: {
+              isDonating: true
+            }
+          });
+
           return {
-            type: 'info',
-            message: 'User is already donating.'
+            isDonating: true
+          } as const;
+        } catch (error) {
+          fastify.log.error(error);
+          void reply.code(500);
+          return {
+            type: 'danger',
+            message: 'Something went wrong.'
           } as const;
         }
+      }
+    );
 
-        await fastify.prisma.user.update({
-          where: { id: req.session.user.id },
-          data: {
-            isDonating: true
+    fastify.post(
+      '/donate/charge-stripe-card',
+      {
+        schema: schemas.chargeStripeCard
+      },
+      async (req, reply) => {
+        try {
+          const { paymentMethodId, amount, duration } = req.body;
+          const { id } = req.session.user;
+
+          const user = await fastify.prisma.user.findUnique({
+            where: { id }
+          });
+
+          if (!user) {
+            void reply.code(500);
+            return {
+              type: 'danger',
+              message: 'Something went wrong.'
+            } as const;
           }
-        });
-
-        return {
-          isDonating: true
-        } as const;
-      } catch (error) {
-        fastify.log.error(error);
-        void reply.code(500);
-        return {
-          type: 'danger',
-          message: 'Something went wrong.'
-        } as const;
+
+          const { email, name } = user;
+
+          if (user.isDonating) {
+            void reply.code(400);
+            return {
+              type: 'info',
+              message: 'User is already donating.'
+            } as const;
+          }
+
+          // Create Stripe Customer
+          let customerId;
+          try {
+            const customer = await stripe.customers.create({
+              email,
+              payment_method: paymentMethodId,
+              invoice_settings: { default_payment_method: paymentMethodId },
+              ...(name && { name })
+            });
+            customerId = customer?.id;
+          } catch {
+            throw {
+              type: 'customerCreationFailed',
+              message: 'Failed to create stripe customer'
+            };
+          }
+
+          // //Create Stripe Subscription
+          let subscriptionId;
+
+          const plan = `${donationSubscriptionConfig.duration[
+            duration
+          ].toLowerCase()}-donation-${amount}`;
+
+          try {
+            const {
+              id: subscription_id,
+              latest_invoice: {
+                // following key is not present in new api definitions. @ts-ignore as recommended by Stripe docs
```

Comment:
Should we add a link to the relevant docs section in this comment? That way we can track + remove when the API def is updated?

Reply:
Definitely the right call.